src/os/bluestore/BitMapAllocator.cc

```diff
@@ -119,14 +90,43 @@ int64_t BitMapAllocator::allocate(
 
   assert(!max_alloc_size || max_alloc_size >= alloc_unit);
 
+  {
+    int nblks = want_size / m_block_size; // apply floor
+    assert(!(want_size % m_block_size));
+    dout(10) << __func__ << " instance " << (uint64_t) this
+           << " num_used " << m_bit_alloc->get_used_blocks()
+           << " total " << m_bit_alloc->total_blocks()
+           << dendl;
+
+    if (!m_bit_alloc->reserve_blocks(nblks)) {
+      return -ENOSPC;
+    }
+  }
+
+
   dout(10) << __func__ <<" instance "<< (uint64_t) this
      << " want_size " << want_size
      << " alloc_unit " << alloc_unit
      << " hint " << hint
      << dendl;
   hint = hint % m_total_size; // make hint error-tolerant
-  return allocate_dis(want_size, alloc_unit / m_block_size,
+  auto res = allocate_dis(want_size, alloc_unit / m_block_size,
                       max_alloc_size, hint / m_block_size, extents);
+
+  if (res < want_size) {
```

Comment:
@ifed01 
This creates a warning now:

```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/BitMapAllocator.cc: In member function 'virtual int64_t BitMapAllocator::allocate(uint64_t, uint64_t, uint64_t, int64_t, PExtentVector*)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/os/bluestore/BitMapAllocator.cc:116:11: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   if (res < want_size) {
       ~~~~^~~~~~~~~~~
```

Reply:
@joscollin : thanks for the note, fixed by https://github.com/ceph/ceph/pull/22064